### DISCUSSION TITLE
Help on soft keyboard handling

### DISCUSSION THREAD

**Discussion by Marekeon** [beeware/toga#2452] [Q&A]
Hi, im still fairly new to toga but making quick progress with my Android app. What I'm struggling with right now is the fact that when I change focus after having performed summer user input in a text fields, the keyboard remains open. 

I saw that there is already an engagement"iOS App needs a "native" way to remove keyboard from view #1236" but it's already quite old and open. Maybe someone had the same issue and found a workaround?
Many thanks in advance!

**Comment by freakboy3742** [comment]
The only workaround I can think of would be to invoke the native API to hide the keyboard - on a text input I believe: 

    my_text_input._impl.native.endEditing(True)

should work; another approach that should work on any input widget would be:

    my_widget._impl.native.resignFirstResponder()

There are iOS-specific APIs though - they won't work on any other platform.

**Reply by Marekeon** [reply]
Thanks for the feedback!
That doesn't seem to work on my android setup indeed.. Is there any way to use this  "_impl.native." to get access to java/kotlin and issue commands around the InputMethodManager ?   or to send a command, like the sendKeys() in some other programming languages? sending the key "back" would probably do the trick as well?

**Reply by freakboy3742** [reply]
The `..._impl.native` part will be the same on every platform; it's what you invoke *on* the native object that will change. 

I don't know the Android solution off the top of my head, but [this SO answer](https://stackoverflow.com/questions/1109022/how-can-i-close-hide-the-android-soft-keyboard-programmatically) has a couple of suggestions. From the look of it, they might not be methods on the native textinput widget, though, so you'll need to find a path to the native widget that has the API.

**Reply by mhsmith** [reply]
In that SO answer:
* The equivalent of `view` is `widget._impl.native`, where `widget` is any widget in your app.
* The equivalent of the implicit `this` on which `getSystemService` is called is `app._impl.native`, where `app` is your `App` object.

**Comment by Marekeon** [comment]
Thank you very much for the hints! My different attempts all miserably failed for now (a bit beyond my competence...), but as a temporary workaround, just reloading the window content worked: 
`self.app.main_window.content = self`

I'll check later if I manage to use this _impl.native, as that really opens infinite doors :D
